discovery_engine_core/core/src/storage/sqlite.rs

```diff
@@ -540,6 +545,57 @@ impl SearchScope for SqliteStorage {
     }
 }
 
+#[async_trait]
+impl StateScope for SqliteStorage {
+    async fn store(&self, bytes: Vec<u8>) -> Result<(), Error> {
+        let mut tx = self.pool.begin().await?;
+
+        sqlx::query("DELETE FROM SerializedState;")
+            .execute(&mut tx)
+            .await
+            .map_err(|err| Error::Database(err.into()))?;
+        sqlx::query("INSERT INTO SerializedState (state) VALUES (?);")
+            .bind(bytes)
+            .execute(&mut tx)
+            .await
+            .map_err(|err| Error::Database(err.into()))?;
+
+        tx.commit().await.map_err(Into::into)
+    }
+
+    async fn fetch(&self) -> Result<Option<Vec<u8>>, Error> {
+        let mut tx = self.pool.begin().await?;
+
+        let state = sqlx::query_as::<_, QueriedState>("SELECT state FROM SerializedState;")
+            .fetch_one(&mut tx)
+            .await
+            .or_else(|err| {
+                if let sqlx::Error::RowNotFound = err {
+                    Ok(QueriedState::default())
+                } else {
+                    Err(Error::Database(err.into()))
+                }
+            })?;
+
+        tx.commit().await?;
+
+        Ok((!state.state.is_empty()).then(|| state.state))
```

Comment:
🎨 we could also use `fetch_optional`
```suggestion
        let state = sqlx::query_as::<_, QueriedState>("SELECT state FROM SerializedState;")
            .fetch_optional(&mut tx)
            .await
            .or_else(|err| {
                if let sqlx::Error::RowNotFound = err {
                    Ok(None)
                } else {
                    Err(Error::Database(err.into()))
                }
            })?;

        tx.commit().await?;

        Ok(state.map(|s| s.state))
```

Reply:
that's better, i'll change it 👍